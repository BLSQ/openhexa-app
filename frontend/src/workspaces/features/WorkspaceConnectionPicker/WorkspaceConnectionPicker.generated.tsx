import * as Types from '../../../graphql/types';

import { gql } from '@apollo/client';
export type WorkspaceConnectionPicker_WorkspaceFragment = { __typename?: 'Workspace', slug: string, connections: Array<{ __typename?: 'CustomConnection', id: string, name: string, slug: string, type: Types.ConnectionType } | { __typename?: 'DHIS2Connection', id: string, name: string, slug: string, type: Types.ConnectionType } | { __typename?: 'GCSConnection', id: string, name: string, slug: string, type: Types.ConnectionType } | { __typename?: 'IASOConnection', id: string, name: string, slug: string, type: Types.ConnectionType } | { __typename?: 'PostgreSQLConnection', id: string, name: string, slug: string, type: Types.ConnectionType } | { __typename?: 'S3Connection', id: string, name: string, slug: string, type: Types.ConnectionType }> };

export const WorkspaceConnectionPicker_WorkspaceFragmentDoc = gql`
    fragment WorkspaceConnectionPicker_workspace on Workspace {
  slug
  connections {
    id
    name
    slug
    type
  }
}
    `;