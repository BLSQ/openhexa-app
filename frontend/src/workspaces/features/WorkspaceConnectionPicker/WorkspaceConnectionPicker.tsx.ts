import { graphql } from "graphql/gql";

export const WorkspaceConnectionPickerWorkspaceDoc = graphql(`
fragment WorkspaceConnectionPicker_workspace on Workspace {
  slug
  connections {
    id
    name
    slug
    type
  }
}
`);
