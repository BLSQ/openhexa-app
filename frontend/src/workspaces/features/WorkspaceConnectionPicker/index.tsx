import WorkspaceConnectionPage from "pages/workspaces/[workspaceSlug]/connections/[connectionId].page";

export default WorkspaceConnectionPage;
