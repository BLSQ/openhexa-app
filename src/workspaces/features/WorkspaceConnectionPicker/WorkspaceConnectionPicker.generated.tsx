import * as Types from '../../../graphql-types';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions = {} as const;
export type WorkspaceConnectionPickerQueryVariables = Types.Exact<{
  slug: Types.Scalars['String']['input'];
}>;


export type WorkspaceConnectionPickerQuery = { __typename?: 'Query', workspace?: { __typename?: 'Workspace', slug: string, connections: Array<{ __typename?: 'Connection', id: string, name: string, slug: string, type: Types.ConnectionType }> } | null };

export type WorkspaceConnectionPicker_WorkspaceFragment = { __typename?: 'Workspace', slug: string, connections: Array<{ __typename?: 'Connection', id: string, name: string, slug: string, type: Types.ConnectionType }> };

export const WorkspaceConnectionPicker_WorkspaceFragmentDoc = gql`
    fragment WorkspaceConnectionPicker_workspace on Workspace {
  slug
  connections {
    id
    name
    slug
    type
  }
}
    `;
export const WorkspaceConnectionPickerDocument = gql`
    query WorkspaceConnectionPicker($slug: String!) {
  workspace(slug: $slug) {
    slug
    ...WorkspaceConnectionPicker_workspace
  }
}
    ${WorkspaceConnectionPicker_WorkspaceFragmentDoc}`;

/**
 * __useWorkspaceConnectionPickerQuery__
 *
 * To run a query within a React component, call `useWorkspaceConnectionPickerQuery` and pass it any options that fit your needs.
 * When your component renders, `useWorkspaceConnectionPickerQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useWorkspaceConnectionPickerQuery({
 *   variables: {
 *      slug: // value for 'slug'
 *   },
 * });
 */
export function useWorkspaceConnectionPickerQuery(baseOptions: Apollo.QueryHookOptions<WorkspaceConnectionPickerQuery, WorkspaceConnectionPickerQueryVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useQuery<WorkspaceConnectionPickerQuery, WorkspaceConnectionPickerQueryVariables>(WorkspaceConnectionPickerDocument, options);
      }
export function useWorkspaceConnectionPickerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<WorkspaceConnectionPickerQuery, WorkspaceConnectionPickerQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useLazyQuery<WorkspaceConnectionPickerQuery, WorkspaceConnectionPickerQueryVariables>(WorkspaceConnectionPickerDocument, options);
        }
export function useWorkspaceConnectionPickerSuspenseQuery(baseOptions?: Apollo.SuspenseQueryHookOptions<WorkspaceConnectionPickerQuery, WorkspaceConnectionPickerQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useSuspenseQuery<WorkspaceConnectionPickerQuery, WorkspaceConnectionPickerQueryVariables>(WorkspaceConnectionPickerDocument, options);
        }
export type WorkspaceConnectionPickerQueryHookResult = ReturnType<typeof useWorkspaceConnectionPickerQuery>;
export type WorkspaceConnectionPickerLazyQueryHookResult = ReturnType<typeof useWorkspaceConnectionPickerLazyQuery>;
export type WorkspaceConnectionPickerSuspenseQueryHookResult = ReturnType<typeof useWorkspaceConnectionPickerSuspenseQuery>;
export type WorkspaceConnectionPickerQueryResult = Apollo.QueryResult<WorkspaceConnectionPickerQuery, WorkspaceConnectionPickerQueryVariables>;