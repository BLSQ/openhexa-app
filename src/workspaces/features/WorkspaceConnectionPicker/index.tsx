import WorkspaceConnectionPage from "pages/workspaces/[workspaceSlug]/connections/[connectionId]";

export default WorkspaceConnectionPage;
