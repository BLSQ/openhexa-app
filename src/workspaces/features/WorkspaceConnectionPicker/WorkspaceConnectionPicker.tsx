import { gql, useQuery } from "@apollo/client";
import { useCallback, useMemo, useState } from "react";
import { useTranslation } from "react-i18next";
import useDebounce from "core/hooks/useDebounce";
import { Combobox } from "core/components/forms/Combobox";
import { ConnectionType } from "graphql-types";
import Connections from "workspaces/helpers/connections";
import { WorkspaceConnectionPickerQuery } from "./WorkspaceConnectionPicker.generated";

type Option = {
  id: string;
  slug: string;
  name: string;
  type: ConnectionType;
};

type WorkspaceConnectionPickerProps = {
  value?: string;
  workspaceSlug: string;
  placeholder?: string;
  onChange(value?: Option): void;
  required?: boolean;
  disabled?: boolean;
  withPortal?: boolean;
  type?: ConnectionType;
};

const WorkspaceConnectionPicker = (props: WorkspaceConnectionPickerProps) => {
  const { t } = useTranslation();
  const {
    workspaceSlug,
    value,
    disabled = false,
    required = false,
    withPortal = false,
    onChange,
    placeholder = t("Select connection"),
    type,
  } = props;

  const { data, loading } = useQuery<WorkspaceConnectionPickerQuery>(
    gql`
      query WorkspaceConnectionPicker($slug: String!) {
        workspace(slug: $slug) {
          slug
          ...WorkspaceConnectionPicker_workspace
        }
      }
      ${WorkspaceConnectionPicker.fragments.workspace}
    `,
    { variables: { slug: workspaceSlug } },
  );

  const [query, setQuery] = useState("");
  const debouncedQuery = useDebounce(query, 150);

  const options = useMemo(() => {
    const lowercaseQuery = debouncedQuery.toLowerCase();
    return (
      data?.workspace?.connections?.filter((c) => {
        if (type) {
          return (
            c.name.toLowerCase().includes(lowercaseQuery) &&
            c.type.toLowerCase() === type.toLowerCase()
          );
        }
        return c.name.toLowerCase().includes(lowercaseQuery);
      }) ?? []
    );
  }, [data, debouncedQuery, type]);

  const displayValue = useCallback(
    (option: Option) => (option ? option.name : ""),
    [],
  );

  const comboBoxValue = useMemo(() => {
    return options.find((option) => option.slug === value) ?? null;
  }, [value, options]);

  return (
    <Combobox
      required={required}
      onChange={onChange}
      loading={loading}
      withPortal={withPortal}
      displayValue={displayValue}
      by="slug"
      onInputChange={useCallback(
        (event: any) => setQuery(event.target.value),
        [],
      )}
      placeholder={placeholder}
      value={comboBoxValue as any}
      onClose={useCallback(() => setQuery(""), [])}
      disabled={disabled}
    >
      {options.map((option: Option) => (
        <Combobox.CheckOption key={option.id} value={option}>
          <div className="flex items-center">
            <img
              loading="lazy"
              src={Connections[option.type as keyof typeof Connections].iconSrc}
              className="sr-hidden mr-2"
              width={16}
              height={11}
              alt="Connection logo"
            />
            {option.name}
          </div>
        </Combobox.CheckOption>
      ))}
    </Combobox>
  );
};

WorkspaceConnectionPicker.fragments = {
  workspace: gql`
    fragment WorkspaceConnectionPicker_workspace on Workspace {
      slug
      connections {
        id
        name
        slug
        type
      }
    }
  `,
};

export default WorkspaceConnectionPicker;
