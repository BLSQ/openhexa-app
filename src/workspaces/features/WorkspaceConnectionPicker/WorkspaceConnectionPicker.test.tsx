import WorkspaceConnectionPicker from "./WorkspaceConnectionPicker";
import { queryByRole, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockedProvider } from "@apollo/client/testing";
import { faker } from "@faker-js/faker";
import { useQuery } from "@apollo/client";
import { ConnectionType } from "graphql-types";

jest.mock("@apollo/client", () => ({
  __esModule: true,
  useQuery: jest.fn(),
  gql: jest.fn(() => "GQL"),
}));

const useQueryMock = useQuery as jest.Mock;

const WORKSPACE = {
  slug: faker.string.uuid(),
  connections: [
    {
      id: faker.string.uuid(),
      name: "dhis2-dev",
      slug: "dhis2-dev",
      type: ConnectionType.Dhis2,
    },
    {
      id: faker.string.uuid(),
      name: "dhis2-staging",
      slug: "dhis2-staging",
      type: ConnectionType.Dhis2,
    },
    {
      id: faker.string.uuid(),
      name: "iaso-dev",
      slug: "iaso-dev",
      type: ConnectionType.Iaso,
    },
  ],
};

describe("WorkspaceConnectionPicker", () => {
  it("display all connections", async () => {
    const user = userEvent.setup();

    useQueryMock.mockReturnValue({
      loading: true,
      data: {
        workspace: WORKSPACE,
      },
    });
    const onChange = jest.fn();

    const { container } = render(
      <WorkspaceConnectionPicker
        workspaceSlug={WORKSPACE.slug}
        onChange={onChange}
        value={""}
      />,
    );

    await user.click(await screen.findByTestId("combobox-button"));
    const option = await screen.queryAllByRole("option");
    expect(option.length).toBe(WORKSPACE.connections.length);

    await user.click(await screen.findByText(WORKSPACE.connections[0].name));
    expect(onChange).toHaveBeenCalledWith(WORKSPACE.connections[0]);
    expect(container).toMatchSnapshot();
  });

  it("display only connections with a given type", async () => {
    const user = userEvent.setup();

    useQueryMock.mockReturnValue({
      loading: true,
      data: {
        workspace: WORKSPACE,
      },
    });
    const onChange = jest.fn();

    render(
      <WorkspaceConnectionPicker
        workspaceSlug={WORKSPACE.slug}
        onChange={onChange}
        value={[]}
        type={ConnectionType.Dhis2}
      />,
    );

    await user.click(await screen.findByTestId("combobox-button"));
    const option = await screen.queryAllByRole("option");
    const dhis2Connections = WORKSPACE.connections.filter(
      (c) => c.type === ConnectionType.Dhis2,
    );

    expect(option.length).toBe(dhis2Connections.length);
  });
});
