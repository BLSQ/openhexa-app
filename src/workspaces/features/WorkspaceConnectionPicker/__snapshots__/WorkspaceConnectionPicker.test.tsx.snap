// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceConnectionPicker display all connections 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      class="form-input flex w-full items-center rounded-md border-gray-300 shadow-sm disabled:border-gray-300 focus-within:outline-none focus:ring-transparent focus-visible:border-blue-500 disabled:cursor-not-allowed  sm:text-sm border-blue-500"
    >
      <div
        class="mr-1 flex flex-1 items-center truncate"
      >
        <input
          aria-activedescendant="headlessui-combobox-option-ee4137b3-3d69-4318-b729-8a2c38c05dd8"
          aria-autocomplete="list"
          aria-controls="headlessui-combobox-options-b608a8f9-e22c-462e-95cb-e9c54e6fa91f"
          aria-expanded="true"
          autocomplete="off"
          class="flex-1 placeholder-gray-600 placeholder-opacity-70 outline-none"
          data-headlessui-state="open"
          data-testid="combobox-input"
          id="headlessui-combobox-input-6b042125-686a-43e0-8a68-23cf5bee102e"
          placeholder="Select connection"
          role="combobox"
          type="text"
        />
      </div>
      <button
        aria-controls="headlessui-combobox-options-b608a8f9-e22c-462e-95cb-e9c54e6fa91f"
        aria-expanded="true"
        aria-haspopup="listbox"
        data-headlessui-state="open"
        data-testid="combobox-button"
        id="headlessui-combobox-button-6f8b5ad0-cfb4-4c17-9e44-20a3470929b1"
        tabindex="-1"
        type="button"
      >
        <div
          class="ml-1 flex items-center gap-0.5 rounded-r-md text-gray-400 focus:outline-none"
        >
          <svg
            class="animate-spin h-5 w-5"
            data-testid="spinner"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              class="opacity-25"
              cx="12"
              cy="12"
              r="10"
              stroke="currentColor"
              stroke-width="4"
            />
            <path
              class="opacity-75"
              d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
              fill="currentColor"
            />
          </svg>
        </div>
      </button>
    </div>
    <ul
      aria-labelledby="headlessui-combobox-button-6f8b5ad0-cfb4-4c17-9e44-20a3470929b1"
      class="max-h-60 z-10 w-full rounded-md bg-white py-1 text-base shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none sm:text-sm flex flex-col"
      data-headlessui-state="open"
      data-popper-escaped="true"
      data-popper-placement="bottom-start"
      data-popper-reference-hidden="true"
      data-testid="combobox-options"
      id="headlessui-combobox-options-b608a8f9-e22c-462e-95cb-e9c54e6fa91f"
      role="listbox"
      style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 4px);"
    >
      <div
        class="overflow-auto flex-1"
        role="none"
      >
        <li
          aria-selected="false"
          class="relative cursor-default select-none px-2 py-2 bg-blue-500 text-white"
          data-headlessui-state="active"
          id="headlessui-combobox-option-ee4137b3-3d69-4318-b729-8a2c38c05dd8"
          role="option"
          tabindex="-1"
        >
          <div
            class="group flex w-full items-center"
          >
            <span
              class="flex items-center pr-4 invisible text-white"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5"
                data-slot="icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m4.5 12.75 6 6 9-13.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
            <span
              class="flex-1 truncate"
            >
              <div
                class="flex items-center"
              >
                <img
                  alt="Connection logo"
                  class="sr-hidden mr-2"
                  height="11"
                  loading="lazy"
                  src="/images/dhis2.svg"
                  width="16"
                />
                dhis2-dev
              </div>
            </span>
          </div>
        </li>
        <li
          aria-selected="false"
          class="relative cursor-default select-none px-2 py-2 text-gray-900"
          data-headlessui-state=""
          id="headlessui-combobox-option-fdd1ab9c-0cb4-44a5-8c7c-f42e8faf8825"
          role="option"
          tabindex="-1"
        >
          <div
            class="group flex w-full items-center"
          >
            <span
              class="flex items-center pr-4 invisible text-gray-900"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5"
                data-slot="icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m4.5 12.75 6 6 9-13.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
            <span
              class="flex-1 truncate"
            >
              <div
                class="flex items-center"
              >
                <img
                  alt="Connection logo"
                  class="sr-hidden mr-2"
                  height="11"
                  loading="lazy"
                  src="/images/dhis2.svg"
                  width="16"
                />
                dhis2-staging
              </div>
            </span>
          </div>
        </li>
        <li
          aria-selected="false"
          class="relative cursor-default select-none px-2 py-2 text-gray-900"
          data-headlessui-state=""
          id="headlessui-combobox-option-2b05dbb3-79f3-434c-8708-3c409d6b88fc"
          role="option"
          tabindex="-1"
        >
          <div
            class="group flex w-full items-center"
          >
            <span
              class="flex items-center pr-4 invisible text-gray-900"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5"
                data-slot="icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m4.5 12.75 6 6 9-13.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
            <span
              class="flex-1 truncate"
            >
              <div
                class="flex items-center"
              >
                <img
                  alt="Connection logo"
                  class="sr-hidden mr-2"
                  height="11"
                  loading="lazy"
                  src="/images/iaso.svg"
                  width="16"
                />
                iaso-dev
              </div>
            </span>
          </div>
        </li>
      </div>
    </ul>
  </div>
</div>
`;
